### PR TITLE
Proper fix utf8 command line arguments

### PR DESCRIPTION
**Pull request by hknielsen**
https://github.com/protocolbuffers/protobuf/pull/14197
Tried to fix utf-8 issue, but it didnt handle multibyte chars.
Only way I found that works constantly is using `CommandLineToArgvW`.
To not ripple out `wchar_t`, I convert to and from where needed

### CODE REVIEW

**Comment by hknielsen** [comment]
Fyi @fowles 
Found that the prev PR didnt work with multi byte chars. This PR fixes that.
To not ripple `wchar_t` out trough the codebase I convert to wstring, ie. when reading the file

File: src/google/protobuf/compiler/command_line_interface.cc [hunk 1]
```diff
@@ -1771,10 +1771,23 @@ bool CommandLineInterface::MakeInputsBeProtoPathRelative(
...
+// On windows to force ifstream to handle proper utr-8, we need to convert to proper supported utf8 wstring.
+// If we dont then the file can't be opened.
+#ifdef _MSC_VER
+  // Convert the file name to wide chars.
+  int size = MultiByteToWideChar(CP_UTF8, 0, file, strlen(file), NULL, 0);
+  std::wstring fileStr;
+  fileStr.resize(size);
+  MultiByteToWideChar(CP_UTF8, 0, file, strlen(file), &fileStr[0], fileStr.size());
+#else
+  std::string fileStr(file);
```
Comment (fowles) [inline src/google/protobuf/compiler/command_line_interface.cc:1785]: nit: file_str to adhere to the google style for variable naming

File: src/google/protobuf/compiler/main.cc [hunk 2]
```diff
@@ -101,6 +105,30 @@ int ProtobufMain(int argc, char* argv[]) {
...
+    int size = WideCharToMultiByte(CP_UTF8, 0, input, wcslen(input), 0, 0, NULL, NULL);
+    std::string result(size, 0);
+    if (size) WideCharToMultiByte(CP_UTF8, 0, input, wcslen(input), &result[0], size, NULL, NULL);
+    return result;
+}
+
+int main(int argc, char* argv[]) 
+{
+  wchar_t** wargv = CommandLineToArgvW(GetCommandLineW(), &argc);
+  std::string* argv_mbcs_str = new std::string [argc];
```
Comment (fowles) [inline src/google/protobuf/compiler/main.cc:120]: you can drop this array and leak each string inside the loop (since we leak all the strings anyway)

**Comment by fowles** [comment]
a bunch of stuff in our repo/ci moved around recently.  Can you rebase this change?

**Comment by fowles** [comment]
@hknielsen pending comments and needs a rebase

**Comment by hlopko** [comment]
Hi @hknielsen, do you plan to continue working on this CL? Thanks!

**Comment by hknielsen** [comment]
@fowles @hlopko sorry for the delay. I pushed a rebase, let me know if I need to do anything more

File: src/google/protobuf/compiler/main.cc (hunk 2 above)
Comment (hknielsen) [reply src/google/protobuf/compiler/main.cc:120]: Do you want me to change to leak so its more simple, or do you think its fine?

File: src/google/protobuf/compiler/command_line_interface.cc (hunk 1 above)
Comment (hknielsen) [reply src/google/protobuf/compiler/command_line_interface.cc]: Fixed https://github.com/protocolbuffers/protobuf/pull/14253/commits/cfbdd357cbed555f8a09d186dbc1f26943002c1c

**Comment by hknielsen** [comment]
@hlopko @fowles are there anything more I need to do to move this PR along :)?

File: src/google/protobuf/compiler/main.cc (hunk 2 above)
Comment (fowles) [reply src/google/protobuf/compiler/main.cc:120]: I would prefer you drop this array to keep the code a bit more simple

Comment (hknielsen) [reply src/google/protobuf/compiler/main.cc]: Fixed in [8e9879b](https://github.com/protocolbuffers/protobuf/pull/14253/commits/8e9879bacfeecedbb5bfee6150d580b89fd55c4e)
Could not test on my windows machine currently, will this be validated by any tests?

File: src/google/protobuf/compiler/main.cc [hunk 3]
```diff
@@ -101,6 +105,29 @@ int ProtobufMain(int argc, char* argv[]) {
...
+}
+
+int main(int argc, char* argv[]) 
+{
+  wchar_t** wargv = CommandLineToArgvW(GetCommandLineW(), &argc);
+  char** argv_mbcs = new char* [argc];
+  for (int i = 0; i < argc; i++)
+  {
+      auto multibyte_string = ToMultiByteUTF8String(wargv[i]);
+      argv_mbcs[i] = const_cast<char*>(multibyte_string.c_str());
```
Comment (fowles) [inline src/google/protobuf/compiler/main.cc:124]: this is now a dangling pointer.  What you wanted here was

```
      std::string* multibyte_string = new auto(ToMultiByteUTF8String(wargv[i]));
      argv_mbcs[i] = const_cast<char*>(multibyte_string->c_str());
```

Comment (hknielsen) [reply src/google/protobuf/compiler/main.cc]: Ah yes good catch. Fixed in force push 881ae8e322747955fe2d2351bb7f2cd5b3561758

File: src/google/protobuf/compiler/main.cc [hunk 4]
```diff
@@ -101,6 +105,29 @@ int ProtobufMain(int argc, char* argv[]) {
...
+}
+
+int main(int argc, char* argv[]) 
+{
+  wchar_t** wargv = CommandLineToArgvW(GetCommandLineW(), &argc);
+  char** argv_mbcs = new char* [argc];
+  for (int i = 0; i < argc; i++)
+  {
+      std::string* multibyte_string = new auto(ToMultiByteUTF8String(wargv[i]));
+      argv_mbcs[i] = const_cast<char*>(multibyte_string->c_str());
```
[the remark this answers is not in the document]
Comment (hknielsen) [reply src/google/protobuf/compiler/main.cc]: Ah yes, much better:
https://github.com/protocolbuffers/protobuf/pull/14253/commits/31b3d257d25af4a6c74e8ffe953a9fea5d37bb22

[the remark this answers is not in the document]
Comment (hknielsen) [reply src/google/protobuf/compiler/main.cc]: @fowles what does this failure mean:
https://github.com/protocolbuffers/protobuf/actions/runs/7387721436/job/20097054069
Both Bazel7 actions seem to fail

File: src/google/protobuf/compiler/main.cc [hunk 5]
```diff
@@ -101,6 +105,29 @@ int ProtobufMain(int argc, char* argv[]) {
 }  // namespace protobuf
 }  // namespace google
 
-int main(int argc, char* argv[]) {
+#ifdef _MSC_VER
+std::string ToMultiByteUTF8String(const wchar_t* input)
```
Comment (fowles) [inline src/google/protobuf/compiler/main.cc:109]: google3 naming convention is `ToMultiByteUtf8String` (note the lowercase on `tf`)

**Review by fowles — COMMENTED** [review]
last nit, otherwise looks good.

File: src/google/protobuf/compiler/main.cc (hunk 5 above)
Comment (hknielsen) [reply src/google/protobuf/compiler/main.cc]: fixed in [28dbcf5](https://github.com/protocolbuffers/protobuf/pull/14253/commits/28dbcf5b7e30243a296f795a5a7c16f51a99f6b8)

**Review by fowles — APPROVED** [review]

**Comment by fowles** [comment]
thanks for all your patience with this PR!  At this point the PR will go through some internal machinations, so you won't see updates here for a few days, but it should (unless something goes wrong) land next week.  If you don't see any motion next week, feel free to ping me!

**Comment by fowles** [comment]
All the windows tests are failing, so I suspect that something is actually wrong in this PR

**Comment by hknielsen** [comment]
> All the windows tests are failing, so I suspect that something is actually wrong in this PR

@fowles hmm it looks like windows are not using the correct compiler flags for c++17, can that be true? I see a bunch of warnings about it:
`Command line warning D9002 : ignoring unknown option '-std=c++17'`

That also explains:
https://github.com/protocolbuffers/protobuf/actions/runs/7432571452/job/20224540608?pr=14253#step:3:736
Since `data()` noneconst was added in C++17 AFAICS

**Comment by fowles** [comment]
yeah, we have to go back to the `const_cast` because we still support C++14 for 1 more year.

**Comment by hknielsen** [comment]
> yeah, we have to go back to the `const_cast` because we still support C++14 for 1 more year.

Reverted back to `const_cast`

File: src/google/protobuf/compiler/main.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (hknielsen) [reply src/google/protobuf/compiler/main.cc:124]: Went back to const_cast as we need to support older version still

**Review by fowles — APPROVED** [review]

**Review by fowles — APPROVED** [review]

**Comment by hknielsen** [comment]
`Tests / Python / MacOS Pure 3.12 (pull_request_target) ` and `feedback/copybara ` are failing, both of these I cant see what the issue is. @fowles are there something here I need to look into?

**Comment by fowles** [comment]
Nothing for you to do.  There was a merge conflict on internal import that I will handle

**Pull request closed without merging**